### PR TITLE
Implement WebCodecs VideoFrame copyTo

### PR DESCRIPTION
**Pull request by youennf**
#### 13105f5f6a4520b29b20c7f2ac44e892b29ba61e
<pre>
Implement WebCodecs VideoFrame copyTo
<a href="https://bugs.webkit.org/show_bug.cgi?id=246210">https://bugs.webkit.org/show_bug.cgi?id=246210</a>
rdar://problem/100893780

Reviewed by Eric Carlson.

Add support for copyTo for RGB, NV12 and I420 formats for macOS/iOS ports.
Make use of libyuv for I420 and direct copy for RGB and NV12.

* LayoutTests/imported/w3c/web-platform-tests/webcodecs/videoFrame-copyTo.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webcodecs/videoFrame-copyTo.any.js:
(makeNV12_4x2):
(promise_test.async t):
* LayoutTests/imported/w3c/web-platform-tests/webcodecs/videoFrame-copyTo.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webcodecs/videoFrame-copyTo.crossOriginIsolated.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webcodecs/videoFrame-copyTo.crossOriginIsolated.https.any.worker-expected.txt:
* Source/ThirdParty/libwebrtc/Configurations/libwebrtc.iOS.exp:
* Source/ThirdParty/libwebrtc/Configurations/libwebrtc.iOSsim.exp:
* Source/ThirdParty/libwebrtc/Configurations/libwebrtc.mac.exp:
* Source/ThirdParty/libwebrtc/Source/webrtc/sdk/WebKit/WebKitUtilities.h:
* Source/ThirdParty/libwebrtc/Source/webrtc/sdk/WebKit/WebKitUtilities.mm:
(webrtc::copyI420BufferToPixelBuffer):
(webrtc::pixelBufferFromI420Buffer):
(webrtc::copyPixelBufferInI420Buffer):
(webrtc::copyBufferToPixelBuffer): Deleted.
* Source/WebCore/Modules/webcodecs/WebCodecsVideoFrameAlgorithms.cpp:
(WebCore::sampleCountPerPixel):
(WebCore::computeLayoutAndAllocationSize):
* Source/WebCore/platform/graphics/cv/VideoFrameCV.mm:
(WebCore::VideoFrame::createRGBA):
(WebCore::VideoFrame::createI420):
(WebCore::copyRGBData):
(WebCore::copyNV12):
(WebCore::VideoFrame::copyTo):

Canonical link: <a href="https://commits.webkit.org/255429@main">https://commits.webkit.org/255429@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/5509ac23d5e859d8b96ed564b91f14d71799b49c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/92530 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/1751 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/23116 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/102280 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/162702 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/96532 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/1748 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/30118 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/84943 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/98449 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/98193 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/1172 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/79030 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/28106 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/83093 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/82785 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/71197 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/36529 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/16717 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/34309 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/17896 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3765 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/38178 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/40506 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/40087 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/37053 "Passed tests") | |

### CODE REVIEW

File: LayoutTests/imported/w3c/web-platform-tests/webcodecs/videoFrame-copyTo.any-expected.txt [hunk 1]
```diff
@@ -1,17 +1,18 @@
...
+PASS Test stride and offset work.
+PASS Test stride and offset with padding.
 PASS Test invalid stride.
 FAIL Test address overflow. assert_throws_js: function "() => frame.allocationSize(options)" did not throw
-FAIL Test codedRect. promise_test: Unhandled rejection with value: object "TypeError: Unable to copy data"
+PASS Test codedRect.
 PASS Test empty rect.
 PASS Test unaligned rect.
-FAIL Test left crop. promise_test: Unhandled rejection with value: object "TypeError: Unable to copy data"
+FAIL Test left crop. assert_equals: buffer contents at index 0 expected 3 but got 1
```
Comment (eric-carlson) [inline LayoutTests/imported/w3c/web-platform-tests/webcodecs/videoFrame-copyTo.any-expected.txt]: Why does this one still fail?

File: LayoutTests/imported/w3c/web-platform-tests/webcodecs/videoFrame-copyTo.any.worker-expected.txt [hunk 2]
```diff
@@ -1,17 +1,18 @@
...
+PASS Test stride and offset work.
+PASS Test stride and offset with padding.
 PASS Test invalid stride.
 FAIL Test address overflow. assert_throws_js: function "() => frame.allocationSize(options)" did not throw
-FAIL Test codedRect. promise_test: Unhandled rejection with value: object "TypeError: Unable to copy data"
+PASS Test codedRect.
 PASS Test empty rect.
 PASS Test unaligned rect.
-FAIL Test left crop. promise_test: Unhandled rejection with value: object "TypeError: Unable to copy data"
+FAIL Test left crop. assert_equals: buffer contents at index 0 expected 3 but got 1
```
Comment (eric-carlson) [inline LayoutTests/imported/w3c/web-platform-tests/webcodecs/videoFrame-copyTo.any.worker-expected.txt]: Ditto

File: Source/WebCore/platform/graphics/cv/VideoFrameCV.mm [hunk 3]
```diff
@@ -180,79 +178,158 @@
...
+    Vector<PlaneLayout> planeLayouts;
+    planeLayouts.append(planeLayout);
+    return planeLayouts;
+}
+
+static Vector<PlaneLayout> copyNV12(Span<uint8_t> span, const ComputedPlaneLayout& spanPlaneLayoutY, const ComputedPlaneLayout& spanPlaneLayoutUV, CVPixelBufferRef pixelBuffer)
+{
+    auto result = CVPixelBufferLockBaseAddress(pixelBuffer, kCVPixelBufferLock_ReadOnly);
+    if (result != kCVReturnSuccess) {
+        RELEASE_LOG_ERROR(WebRTC, "VideoFrame::copyTo lock failed");
```
Comment (eric-carlson) [inline Source/WebCore/platform/graphics/cv/VideoFrameCV.mm]: Log the error code?

**Review by eric-carlson — APPROVED** [review]
r=me

File: Source/WebCore/platform/graphics/cv/VideoFrameCV.mm (hunk 3 above)
Comment (youennf) [reply Source/WebCore/platform/graphics/cv/VideoFrameCV.mm]: Will do

File: LayoutTests/imported/w3c/web-platform-tests/webcodecs/videoFrame-copyTo.any-expected.txt (hunk 1 above)
Comment (youennf) [reply LayoutTests/imported/w3c/web-platform-tests/webcodecs/videoFrame-copyTo.any-expected.txt]: Only full frame copies are supported right now.
A follow-up should fix this.

File: LayoutTests/imported/w3c/web-platform-tests/webcodecs/videoFrame-copyTo.any.worker-expected.txt (hunk 2 above)
Comment (youennf) [reply LayoutTests/imported/w3c/web-platform-tests/webcodecs/videoFrame-copyTo.any.worker-expected.txt]: Only full frame copies are supported right now.
A follow-up should fix this.

File: LayoutTests/imported/w3c/web-platform-tests/webcodecs/videoFrame-copyTo.any-expected.txt (hunk 1 above)
Comment (youennf) [reply LayoutTests/imported/w3c/web-platform-tests/webcodecs/videoFrame-copyTo.any-expected.txt]: Filed https://bugs.webkit.org/show_bug.cgi?id=246393

**Pull request closed without merging**